src/test/librbd/test_Groups.cc

```diff
@@ -480,3 +480,255 @@ TEST_F(TestGroup, add_snapshotPP)
+
+  const char *image_name1=m_image_name.c_str();
+
+  const char *image_name2="snap_list2PP_image2";
+  ASSERT_EQ(0, create_image_pp(m_rbd, ioctx2, image_name2, m_image_size));
+
+  const char *gp_name= "snap_group";
+  ASSERT_EQ(0, m_rbd.group_create(m_ioctx, gp_name));
+
+  std::vector<librbd::group_snap_info2_t> gp_snaps;
+  ASSERT_EQ(-ENOENT, m_rbd.group_snap_list2(m_ioctx, "absent", &gp_snaps));
+  ASSERT_EQ(0, m_rbd.group_snap_list2(m_ioctx, gp_name, &gp_snaps));
+  ASSERT_EQ(0U, gp_snaps.size());
+
+  const char* const gp_snap_names[] = {
+    "snap_snapshot0", "snap_snapshot1", "snap_snapshot2", "snap_snapshot3"};
+
+  ASSERT_EQ(0, m_rbd.group_snap_create(m_ioctx, gp_name, gp_snap_names[0]));
+
+  ASSERT_EQ(0, m_rbd.group_image_add(m_ioctx, gp_name, m_ioctx,
+                                     image_name1));
+  ASSERT_EQ(0, m_rbd.group_snap_create(m_ioctx, gp_name, gp_snap_names[1]));
+
+  ASSERT_EQ(0, m_rbd.group_image_add(m_ioctx, gp_name, ioctx2,
+                                     image_name2));
+  ASSERT_EQ(0, m_rbd.group_snap_create(m_ioctx, gp_name, gp_snap_names[2]));
+
+  ASSERT_EQ(0, m_rbd.group_image_remove(m_ioctx, gp_name,
+                                        m_ioctx, image_name1));
+  ASSERT_EQ(0, m_rbd.group_snap_create(m_ioctx, gp_name, gp_snap_names[3]));
+
+  ASSERT_EQ(0, m_rbd.group_snap_list2(m_ioctx, gp_name, &gp_snaps));
+  ASSERT_EQ(4U, gp_snaps.size());
+
+  for (const auto& gp_snap : gp_snaps) {
+    if (gp_snap.name == gp_snap_names[0]) {
+      ASSERT_EQ(0U, gp_snap.image_snaps.size());
+    } else if (gp_snap.name == gp_snap_names[1]) {
+      ASSERT_EQ(1U, gp_snap.image_snaps.size());
+      ASSERT_EQ(image_name1, gp_snap.image_snaps[0].image_name);
+      ASSERT_EQ(m_ioctx.get_id(), gp_snap.image_snaps[0].pool_id);
+    } else if (gp_snap.name == gp_snap_names[2]) {
+      ASSERT_EQ(2U, gp_snap.image_snaps.size());
+      for (int j = 0; j < 2; j++) {
+	if (gp_snap.image_snaps[j].image_name == image_name1) {
+	  ASSERT_EQ(m_ioctx.get_id(), gp_snap.image_snaps[j].pool_id);
+	} else if (gp_snap.image_snaps[j].image_name == image_name2) {
+	  ASSERT_EQ(ioctx2.get_id(), gp_snap.image_snaps[j].pool_id);
+	} else {
+          FAIL() << "Unexpected image in group snap: "
+                 << gp_snap.image_snaps[j].image_name;
+	}
+      }
+    } else if(gp_snap.name == gp_snap_names[3]) {
+      ASSERT_EQ(1U, gp_snap.image_snaps.size());
+      ASSERT_EQ(image_name2, gp_snap.image_snaps[0].image_name);
+      ASSERT_EQ(ioctx2.get_id(), gp_snap.image_snaps[0].pool_id);
+    } else {
+      FAIL() << "Unexpected group snap: " << gp_snap.name;
+    }
+  }
+
+  for (const auto& gp_snap_name : gp_snap_names) {
+    ASSERT_EQ(0, m_rbd.group_snap_remove(m_ioctx, gp_name, gp_snap_name));
+  }
+  std::vector<librbd::group_snap_info2_t> gp_snaps2;
+  ASSERT_EQ(0, m_rbd.group_snap_list2(m_ioctx, gp_name, &gp_snaps2));
+  ASSERT_EQ(0U, gp_snaps2.size());
```

Comment:
Add a corresponding "listing is now empty" case to the C API test for consistency.